File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P02.scala
package com.shekhargulati.ninetynine_problems._01_lists

object P02 {

  def secondLast[T](list: List[T]): T = list match {
    case x :: (_ :: Nil) => x
    case x :: xs => secondLast(xs)
    case _ => throw new NoSuchElementException("Can't find secondLast element from a list with less than 2 elements")
  }

  def secondLast1[T](list: List[T]): T = if (list.isEmpty) throw new NoSuchElementException else list.init.last

}

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P02.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.LinkedList;
import java.util.List;
import java.util.NoSuchElementException;

public class P02 {

    public static <T> T secondLast(List<T> list) {
        if (list.size() < 2) {
            throw new NoSuchElementException("Can't find secondLast element from a list with less than 2 elements");
        }
        return list.get(list.size() - 2);
    }

    public static <T> T secondLastRecursion(LinkedList<T> list) {
        if (list.size() < 2) {
            throw new NoSuchElementException("Can't find secondLast element from a list with less than 2 elements");
        }
        if (list.size() == 2) {
            return list.getFirst();
        }
        return secondLastRecursion(new LinkedList<>(list.subList(1, list.size())));

    }
}
